.github/workflows/tests.yml

```diff
@@ -0,0 +1,93 @@
+name: Run tests
+permissions:
+  contents: read
+
+on:
+  push:
+    branches:
+      - develop
+      - feature/full-solution
+      - main
+
+  pull_request:
+    types:
+      - opened
+      - reopened
+      - synchronize
+      - ready_for_review
+
+jobs:
+  code_format:
+    name: Check code formatting
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v5
+        with:
+          fetch-depth: 0
+
+      - name: Install uv
+        uses: astral-sh/setup-uv@v7.1.2
+        with:
+          enable-cache: true
+
+      - name: Install dependencies
+        run: uv sync
+
+      - name: Run code formatting checks
+        shell: bash -l {0}
+        continue-on-error: false
+        run: uv run ruff format --check
+
+  code_linting:
+    name: Run code lint checks
+    runs-on: ubuntu-latest
+    needs: code_format
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v5
+        with:
+          fetch-depth: 0
+
+      - name: Install uv
+        uses: astral-sh/setup-uv@v7.1.2
+        with:
+          enable-cache: true
+
+      - name: Install dependencies
+        run: uv sync
+
+      - name: Run code linting checks
+        run: uv run ruff check --output-format=github .
+
+  tests:
+    name: Run local tests (${{ matrix.python-version }}, ${{ matrix.os }})
+    runs-on: ${{ matrix.os }}
+    needs: code_format
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.13"]
+        os: ["ubuntu-24.04"]
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v5
+        with:
+          fetch-depth: 0
+
+      - name: Install uv
+        uses: astral-sh/setup-uv@v7.1.2
+
+      - name: Set up Python
+        uses: actions/setup-python@v6
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install dependencies
+        run: uv sync --dev
+
+      - name: Run local tests
+        shell: bash -l {0}
+        continue-on-error: false
+        run: uv run pytest -m "not requires_spark"
```

Comment:
You can easily add codecov like this:
https://github.com/kbase/cdm-task-service/blob/main/.github/workflows/test.yml#L88-L92